### PR TITLE
Enable index tracking for tools - openai api support

### PR DESCRIPTION
**Pull request by ParthSareen**
Closes https://github.com/ollama/ollama/issues/7881
Now able to use `client.beta.chat.completions.stream`
image

### CODE REVIEW

File: openai/openai_test.go [hunk 1]
```diff
@@ -195,7 +195,86 @@ func TestChatMiddleware(t *testing.T) {
 				Stream: &False,
 			},
 		},
-
+		{
+			name: "chat handler with streaming tools",
```
Comment (ParthSareen) [inline openai/openai_test.go:199]: Quick test to make sure middleware is catching streamed tools (nothing for this PR but just general tracking regressions)

File: server/routes.go [hunk 2]
```diff
@@ -1469,7 +1469,7 @@ func (s *Server) ChatHandler(c *gin.Context) {
 	go func() {
 		defer close(ch)
 		var sb strings.Builder
-		var hasToolCalls bool
+		var toolCallIndex int = -1
```
Comment (jmorganca) [inline server/routes.go:1472]:
```suggestion
		var toolCallIndex int = 0
```

File: server/routes.go [hunk 3]
```diff
@@ -1509,16 +1509,19 @@ func (s *Server) ChatHandler(c *gin.Context) {
 			sb.WriteString(r.Content)
 			if toolCalls, ok := m.parseToolCalls(sb.String()); ok {
 				res.Message.ToolCalls = toolCalls
+				for i := range toolCalls {
+					toolCallIndex++
+					toolCalls[i].Function.Index = toolCallIndex
```
Comment (jmorganca) [inline server/routes.go:1514]:
```suggestion
					toolCalls[i].Function.Index = toolCallIndex
					toolCallIndex++
```

File: server/routes.go [hunk 4]
```diff
@@ -1509,16 +1509,19 @@ func (s *Server) ChatHandler(c *gin.Context) {
...
 				sb.Reset()
-				hasToolCalls = true
 				ch <- res
 				return
 			}
 
 			if r.Done {
 				// Send any remaining content if no tool calls were detected
-				if !hasToolCalls {
+				if toolCallIndex == -1 {
```
Comment (jmorganca) [inline server/routes.go:1524]:
```suggestion
				if toolCalls > 0 {
```

**Review by jmorganca — APPROVED** [review]

**Review by jmorganca — APPROVED** [review]

**Comment by mathisxy** [comment]
When will this be available in the official ollama version?

I still have the problem on version 0.15.4

Im using: ```await self.client.chat.completions.create``` with ```stream=True``` but I just confirmed that its the same with ```await self.client.chat.completions.stream```

Received tool call delta: 
[ChoiceDeltaToolCall(index=0, id='call_juwdh89s', function=ChoiceDeltaToolCallFunction(arguments='{"sides":6}', name='role_dice'), type='function')]
Received tool call delta: 
[ChoiceDeltaToolCall(index=0, id='call_itcm7qna', function=ChoiceDeltaToolCallFunction(arguments='{"sides":6}', name='role_dice'), type='function')]

**Pull request merged**